### PR TITLE
Identify RHEL based instances in AWS.

### PR DESCRIPTION
**Pull request by roliveri**
Information collected and passed to the parser. Once the database and ingress API are updated, this can be merged.

Depends on new GEM version based on:
https://github.com/RedHatInsights/topological_inventory-ingress_api-client-ruby/pull/71

### CODE REVIEW

**Review by lindgrenj6 — COMMENTED** [review]
Just some more stylistic questions. Happy to see this is even _possible_.

File: lib/topological_inventory/amazon/collector/ec2.rb [hunk 1]
```diff
@@ -10,8 +10,31 @@ def source_regions(scope)
...
+
+        def rhel_images_for_instances(ec2, instances)
+          ec2.images(
+            :image_ids => instances.collect(&:image_id).uniq!,
+            :filters   => [
+              {
+                :name   => "platform-details",
+                :values => [
+                  "Red Hat Enterprise Linux",
+                  "Red Hat BYOL Linux"
```
[the remark this answers is not in the document]
Comment (roliveri) [reply lib/topological_inventory/amazon/collector/ec2.rb]: Yes, returns images whose `platform-details` attribute matches either of those strings.

File: lib/topological_inventory/amazon/parser/vm.rb [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
-      def parse_vms(instance, scope)
+      def parse_vms(instance_hash, scope)
+        instance = instance_hash[:instance]
         uid      = instance.id
         name     = get_from_tags(instance.tags, :name) || uid
         flavor   = lazy_find(:flavors, :source_ref => instance.instance_type) if instance.instance_type
         stack_id = get_from_tags(instance.tags, "aws:cloudformation:stack-id")
         stack    = lazy_find(:orchestration_stacks, :source_ref => stack_id) if stack_id
 
+        puts "**** #{uid} => #{instance_hash[:is_rhel]}"
```
[the remark this answers is not in the document]
Comment (roliveri) [reply lib/topological_inventory/amazon/parser/vm.rb:12]: Right. So we'll have to update the table and API to accept the new value.

File: lib/topological_inventory/amazon/collector/ec2.rb [hunk 3]
```diff
@@ -10,8 +10,31 @@ def source_regions(scope)
...
         def vms(scope)
-          ec2_connection(scope).instances
+          ec2 = ec2_connection(scope)
+          instances = ec2.instances
+
+          rhel_images = rhel_images_for_instances(ec2, instances).each_with_object(Hash.new("")) { |img, h| h[img.image_id] = img.platform_details }
+
+          instances.each_with_object([]) do |inst, ary|
+            ary << {:instance => inst, :is_rhel => rhel_images[inst.image_id]}
+          end
```
[the remark this answers is not in the document]
Comment (roliveri) [reply lib/topological_inventory/amazon/collector/ec2.rb]: Updated.

**Review by slemrmartin — COMMENTED** [review]
I put my comment into resolved conversation, so once again:
FYI steps:
- db migration in core
- core pr + core changelog pr (new patch version)
- rake openapi:generate in persister (new column is detected automatically)
  - ingress-api pr
- bump core version in persister (pr)
- rake client:generate in ingress-api
  - update version in client's version.rb (generator makes incorrect version)
  - ingress-api-client-ruby pr
- bump ingress-api-client-ruby version in Amazon's Gemfile
- bump core version in topological-inventory-api (pr)
  - rake openapi:generate in topological-inventory-api
  - rake client:generate in topological-inventory-api (update version.rb in client)
  - topological-inventory-api-client-ruby PR

**Comment by roliveri** [comment]
Thanks Martin, that's very helpful.

**Comment by roliveri** [comment]
@slemrmartin @lindgrenj6 @syncrou - Requesting your opinion.

While looking into the DB and API changes for this, I tried to come up with a name for the new attribute we need to introduce. Initially, this was to be a Boolean, so something like `is_rhel` would be appropriate.

Now however, we also want to distinguish if marketplace or self licensed. To accommodate this, I was planning on making the new attribute a string, and saving the relevant values from the `platform_details` attribute of the instance's image, namely: "Red Hat Enterprise Linux" or "Red Hat BYOL Linux". Here, I'm thinking `rhel_status` or something of the sort might be a good name.

Now I'm wondering if we should just call it `platform_details` and save the value we get from AWS whatever it may be, as it would subsume the RHEL values. While we'll be collecting more information, it may make it more difficult to normalize the information when we add support for Azure and other cloud platforms.

Currently, the AWS `platform_details` attribute can contain the following:
- Linux/UNIX
- Red Hat BYOL Linux
- Red Hat Enterprise Linux
- SQL Server Enterprise
- SQL Server Standard
- SQL Server Web
- SUSE Linux
- Windows
- Windows BYOL
- Windows with SQL Server Enterprise
- Windows with SQL Server Standard
- Windows with SQL Server Web

Thoughts?

**Comment by syncrou** [comment]
@roliveri I like the idea of the more generic name to ease normalization later on and your suggestion of 'platform_details' makes sense. Obviously would like to hear @slemrmartin and @lindgrenj6's thoughts as well.

**Comment by slemrmartin** [comment]
question is, what the 'platform_details' will mean for azure (and future source types). 
From your list it seems the name can be also 'operating_system' which is more self explanatory

**Comment by roliveri** [comment]
Actually, `operating_system` is too narrow. The list includes information on the OS, licensing and software.
I'm thinking `guest_information` might be more appropriate.

Then the question remains, should we save all the listed values, or just the RHEL specific ones? If we save all of them, we may not be able to map Azure information to the same strings. That might not matter. Maybe we should just save what we get.

**Comment by roliveri** [comment]
This should be ready to merge.

**Review by syncrou — APPROVED** [review]
LGTM 👍

**Comment by roliveri** [comment]
I'm a bit confused by the test failure, it's saying:
```
expected collection contained:  [{:flavor=>{:inventory_collection_name=>:flavors, :ref=>:manager_ref, :reference=>{:source_ref=>"m3.m...nager_ref, :reference=>{:source_ref=>"us-west-1"}}, :subscription=>nil, :uid_ems=>"instance_ec2_0"}]

actual collection contained:    [{:flavor=>{:inventory_collection_name=>:flavors, :ref=>:manager_ref, :reference=>{:source_ref=>"m3.m...nager_ref, :reference=>{:source_ref=>"us-west-1"}}, :subscription=>nil, :uid_ems=>"instance_ec2_0"}]

the missing elements were:      [{:flavor=>{:inventory_collection_name=>:flavors, :ref=>:manager_ref, :reference=>{:source_ref=>"m3.m...nager_ref, :reference=>{:source_ref=>"us-west-1"}}, :subscription=>nil, :uid_ems=>"instance_ec2_0"}]
       
the extra elements were:        [{:flavor=>{:inventory_collection_name=>:flavors, :ref=>:manager_ref, :reference=>{:source_ref=>"m3.m...nager_ref, :reference=>{:source_ref=>"us-west-1"}}, :subscription=>nil, :uid_ems=>"instance_ec2_0"}]
```
But what's **expected**, **actual**, **missing** and **extra** are exactly the same. Am I missing something?

**Comment by lindgrenj6** [comment]
@roliveri Running it locally I can get the same failure, looking at the hash you probably need to add the `:guest_info => nil` field to each hash. Diff here:

```diff
diff --git a/spec/collector_spec.rb b/spec/collector_spec.rb
index 101f62c..11e206d 100644
--- a/spec/collector_spec.rb
+++ b/spec/collector_spec.rb
@@ -22,6 +22,7 @@ RSpec.describe TopologicalInventory::Amazon::Collector do
            :name                => "instance_0",
            :orchestration_stack => nil,
            :power_state         => "on",
+           :guest_info          => nil,
            :source_ref          => "instance_0",
            :source_region       =>
                                    {:inventory_collection_name => :source_regions,
@@ -37,6 +38,7 @@ RSpec.describe TopologicalInventory::Amazon::Collector do
            :name                => "instance_ec2_0",
            :orchestration_stack => nil,
            :power_state         => "on",
+           :guest_info          => nil,
            :source_ref          => "instance_ec2_0",
            :source_region       =>
                                    {:inventory_collection_name => :source_regions,
@@ -52,6 +54,7 @@ RSpec.describe TopologicalInventory::Amazon::Collector do
            :name                => "instance_0",
            :orchestration_stack => nil,
            :power_state         => "on",
+           :guest_info          => nil,
            :source_ref          => "instance_0",
            :source_region       =>
                                    {:inventory_collection_name => :source_regions,
@@ -67,6 +70,7 @@ RSpec.describe TopologicalInventory::Amazon::Collector do
            :name                => "instance_ec2_0",
            :orchestration_stack => nil,
            :power_state         => "on",
+           :guest_info          => nil,
            :source_ref          => "instance_ec2_0",
            :source_region       =>
                                    {:inventory_collection_name => :source_regions,

```

**Comment by roliveri** [comment]
@lindgrenj6 That's what I thought. I was confused Because I expected to see `: guest_info` listed as an extra element.

Thanks!

**Comment by roliveri** [comment]
This PR should be ready to merge.

**Review by lindgrenj6 — APPROVED** [review]

**Pull request merged**